src/events/deposits.upsert.service.ts

```diff
@@ -288,4 +229,153 @@ export class DepositsUpsertService {
+    );
+  }
+
+  async upsertWithClient(
+    prisma: BasePrismaClient,
+    operation: UpsertDepositsOperationDto,
+  ): Promise<Deposit[]> {
+    const networkVersion = this.config.get<number>('NETWORK_VERSION');
+    const blockHash = standardizeHash(operation.block.hash);
+
+    const shouldUpsertDeposit =
+      operation.type === BlockOperation.CONNECTED ||
+      operation.type === BlockOperation.DISCONNECTED;
+
+    let deposits = new Array<Deposit>();
+
+    if (shouldUpsertDeposit) {
+      for (const transaction of operation.transactions) {
+        if (operation.type === BlockOperation.CONNECTED) {
+          const amounts = new Map<string, number>();
+
+          for (const note of transaction.notes) {
+            if (note.memo) {
+              const amount = amounts.get(note.memo) ?? 0;
+              amounts.set(note.memo, amount + note.amount);
+            }
+          }
+
+          // Bulk load unique users and map by graffiti
+          const users = await this.usersService.findManyAndMapByGraffiti(
+            Array.from(amounts.keys()),
+          );
+
+          // Create deposit params for each deposit with a matching user
+          const depositParams = new Array<{
+            transaction_hash: string;
+            block_hash: string;
+            block_sequence: number;
+            network_version: number;
+            graffiti: string;
+            main: boolean;
+            amount: number;
+          }>();
+
+          for (const [graffiti, amount] of amounts) {
+            // User deposited with an unknown graffiti
+            if (!users.has(graffiti)) {
+              continue;
+            }
+
+            depositParams.push({
+              graffiti,
+              amount,
+              transaction_hash: standardizeHash(transaction.hash),
+              block_hash: blockHash,
+              block_sequence: operation.block.sequence,
+              main: true,
+              network_version: networkVersion,
+            });
+          }
+
+          // Deposits are shared between blocks, so we need to reassign all the ones on other blocks
+          await prisma.deposit.updateMany({
+            data: {
+              block_hash: blockHash,
+              main: true,
+            },
+            where: {
+              AND: depositParams.map((deposit) => ({
+                transaction_hash: deposit.transaction_hash,
+                graffiti: deposit.graffiti,
+              })),
+              network_version: networkVersion,
+            },
+          });
+
+          // Now create new not existing deposits
+          await prisma.deposit.createMany({
+            data: depositParams,
+            skipDuplicates: true,
+          });
+
+          deposits = await prisma.deposit.findMany({
+            where: {
+              block_hash: blockHash,
+              network_version: networkVersion,
+            },
+          });
+
+          const eventPayloads = deposits.map((deposit) => {
```

Comment:
Doubt that the number of loop iterations matter here, but a friend once told me about how bad `map` is in terms of performance vs `for` loop. Makes me sad since I like syntax/paradigm of `map` better.

https://leanylabs.com/blog/js-forEach-map-reduce-vs-for-for_of/